File: Source/Aladdin.CAPI.PKCS11.Athena/Provider.cs
using System;
using Aladdin.PKCS11; 

namespace Aladdin.CAPI.PKCS11.Athena
{
	///////////////////////////////////////////////////////////////////////////
	// Криптографический провайдер
	///////////////////////////////////////////////////////////////////////////
	public sealed class Provider : ANSI.PKCS11.Provider
	{
		// конструктор
		public Provider() : base("Athena PKCS11 Cryptographic Provider", false) 
        {
            // указать интерфейс вызова функций
            module = Module.Create(new NativeMethods.NativeAPI()); 
        }
        // деструктор
        protected override void OnDispose() 
        {
            // освободить выделенные ресурсы
            RefObject.Release(module); base.OnDispose(); 
        } 
		// интерфейс вызова функций
		public override Module Module { get { return module; }} private Module module;

        // корректная реализация отдельных OAEP механизмов
        public override bool UseOAEP(Applet applet, 
            ANSI.PKCS11.Parameters.CK_RSA_PKCS_OAEP_PARAMS parameters) 
        { 
            // проверить корректность реализации
            return (parameters.HashAlg == API.CKM_SHA_1 && parameters.MGF == API.CKG_MGF1_SHA1); 
        }
        // некорректная реализация PSS механизмов
        public override bool UsePSS(Applet applet, 
            ANSI.PKCS11.Parameters.CK_RSA_PKCS_PSS_PARAMS parameters) { return false; }

	    // создать алгоритм для параметров
	    protected override IAlgorithm CreateAlgorithm(CAPI.Factory factory, 
            SecurityStore scope, string oid, ASN1.IEncodable parameters, Type type)
        {
            // для алгоритмов согласования общего ключа
            if (type == typeof(IKeyAgreement))
            {
                // указать неподдерживаемые алгоритмы
                if (oid == ASN1.ANSI.OID.x962_ec_public_key) return null; 
            }
            // для алгоритмов согласования общего ключа
            else if (type == typeof(ITransportAgreement))
            {
                // указать неподдерживаемые алгоритмы
                if (oid == ASN1.ANSI.OID.x963_ecdh_std_sha1             ) return null; 
                if (oid == ASN1.ANSI.OID.certicom_ecdh_std_sha2_224     ) return null; 
                if (oid == ASN1.ANSI.OID.certicom_ecdh_std_sha2_256     ) return null; 
                if (oid == ASN1.ANSI.OID.certicom_ecdh_std_sha2_384     ) return null; 
                if (oid == ASN1.ANSI.OID.certicom_ecdh_std_sha2_512     ) return null; 
                if (oid == ASN1.ANSI.OID.x963_ecdh_cofactor_sha1        ) return null; 
                if (oid == ASN1.ANSI.OID.certicom_ecdh_cofactor_sha2_224) return null; 
                if (oid == ASN1.ANSI.OID.certicom_ecdh_cofactor_sha2_256) return null; 
                if (oid == ASN1.ANSI.OID.certicom_ecdh_cofactor_sha2_384) return null; 
                if (oid == ASN1.ANSI.OID.certicom_ecdh_cofactor_sha2_512) return null; 
            }
            // вызвать базовую функцию
            return base.CreateAlgorithm(factory, scope, oid, parameters, type); 
        }
    }
}

File: Source/Aladdin.CAPI.PKCS11.Athena.Java/src/aladdin/capi/pkcs11/athena/Provider.java
package aladdin.capi.pkcs11.athena;
import aladdin.asn1.*; 
import aladdin.asn1.ansi.*; 
import aladdin.capi.*; 
import aladdin.capi.pkcs11.*; 
import aladdin.pkcs11.*; 
import aladdin.pkcs11.jni.*;
import java.io.*; 

///////////////////////////////////////////////////////////////////////////
// Криптографический провайдер
///////////////////////////////////////////////////////////////////////////
public class Provider extends aladdin.capi.ansi.pkcs11.Provider
{
    // интерфейс вызова функций
    private final Module module; 
    
	// конструктор
	public Provider(String path) throws IOException 
    { 
        // сохранить переданные параметры
        super("Athena PKCS11 Cryptographic Provider", true);
        
        // открыть модуль
        module = new Module(path); 
    }
	@Override protected void onClose() throws IOException  
    { 
        // освободить выделенные ресурсы
        module.close(); super.onClose();
    } 	
    // интерфейс вызова функций
	@Override public Module module() { return module; } 
    
    // корректная реализация отдельных OAEP механизмов
    @Override public boolean useOAEP(Applet applet, CK_RSA_PKCS_OAEP_PARAMS parameters) 
    { 
        // проверить корректность реализации
        return (parameters.hashAlg == API.CKM_SHA_1 && parameters.mgf == API.CKG_MGF1_SHA1); 
    } 
    // некорректная реализация PSS механизмов
    @Override public boolean usePSS (Applet applet, CK_RSA_PKCS_PSS_PARAMS parameters) { return false; } 
    
	// создать алгоритм генерации ключей
    @Override
	protected aladdin.capi.KeyPairGenerator createGenerator(
        Factory factory, SecurityObject scope, IRand rand, 
        String keyOID, IParameters parameters) throws IOException
    {
        // проверить тип параметров
        if (keyOID.equals(aladdin.asn1.ansi.OID.X962_EC_PUBLIC_KEY))
        {
            // преобразовать тип параметров
            aladdin.capi.ansi.x962.Parameters ecParameters = 
                (aladdin.capi.ansi.x962.Parameters)parameters;

            // найти подходящую смарт-карту
            try (Applet applet = findApplet(scope, API.CKM_EC_KEY_PAIR_GEN, 0, 0))
            {
                // проверить наличие смарт-карты
                if (applet == null) return null; 

                // создать алгоритм генерации ключей
                return new aladdin.capi.pkcs11.athena.x962.KeyPairGenerator(
                    applet, scope, rand, ecParameters
                ); 
            }
        }
        // создать алгоритм генерации ключей
        return super.createGenerator(factory, scope, rand, keyOID, parameters); 
    }
    // создать алгоритм для параметров
    @Override protected IAlgorithm createAlgorithm(Factory factory, 
        SecurityStore scope, String oid, IEncodable parameters, 
        Class<? extends IAlgorithm> type) throws IOException
    {
        // для алгоритмов согласования общего ключа
        if (type.equals(IKeyAgreement.class))
        {
            // указать неподдерживаемые алгоритмы
            if (oid.equals(OID.X962_EC_PUBLIC_KEY)) return null; 
        }
        // для алгоритмов согласования общего ключа
        else if (type.equals(ITransportAgreement.class))
        {
            // указать неподдерживаемые алгоритмы
            if (oid.equals(OID.X963_ECDH_STD_SHA1             )) return null; 
            if (oid.equals(OID.CERTICOM_ECDH_STD_SHA2_224     )) return null; 
            if (oid.equals(OID.CERTICOM_ECDH_STD_SHA2_256     )) return null; 
            if (oid.equals(OID.CERTICOM_ECDH_STD_SHA2_384     )) return null; 
            if (oid.equals(OID.CERTICOM_ECDH_STD_SHA2_512     )) return null; 
            if (oid.equals(OID.X963_ECDH_COFACTOR_SHA1        )) return null; 
            if (oid.equals(OID.CERTICOM_ECDH_COFACTOR_SHA2_224)) return null; 
            if (oid.equals(OID.CERTICOM_ECDH_COFACTOR_SHA2_256)) return null; 
            if (oid.equals(OID.CERTICOM_ECDH_COFACTOR_SHA2_384)) return null; 
            if (oid.equals(OID.CERTICOM_ECDH_COFACTOR_SHA2_512)) return null; 
        }
        // вызвать базовую функцию
        return super.createAlgorithm(factory, scope, oid, parameters, type); 
    }
}
